### DISCUSSION TITLE
/the-basics/validation.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#85] [Announcements]
# /the-basics/validation.html

[[toc]] 简介 Goravel 提供了几种不同的方法来验证传入应用程序的数据。最常见的做法是在所有传入的 HTTP 请求中使用 validate 方法。Goravel 包含了各种方便的验证规则。 快速验证 为了了解 Goravel 强大的验证功能，我们来看一个表单验证并将错误消息展示给用户的完整示例。通过阅读概述，这将会对您如何使用 Gorave...

https://www.goravel.dev/zh/the-basics/validation.html

**Comment by AnkoGo123** [comment]
作为一个中国人，居然不来一个快捷方式的验证器验证是否是中文，真的是gou都不用这个框架

**Comment by xx1196** [comment]
how i can use the unique rule like laravel?

**Reply by hwbrzzl** [reply]
You can use the custom rule for the time being: https://www.goravel.dev/the-basics/validation.html#custom-validation-rules

**Reply by xx1196** [reply]
Is it planned to be implemented natively like Laravel?

For this custom rule, I need make a query, in this case need use gorm? O how I can make this form any model?

**Reply by hwbrzzl** [reply]
Yes, will implement it in the future, for now, you can use `facades.Orm()` in the custom rule directly.

**Comment by xx1196** [comment]
hi, how i can get wich field is in validation? 

return map[string]string{
		"code":     "required|max_len:4|unique:countries",
}


inside on my custom rule, how i can get the attribute code? my rule will be generic for any field

**Reply by hwbrzzl** [reply]
The `val`? https://www.goravel.dev/the-basics/validation.html#custom-validation-rules

<img width="760" alt="image" src="https://github.com/goravel/goravel/assets/24771476/0fdfa944-d701-4eac-9134-ecc5944f47af">

**Reply by 06mrx** [reply]
```go
package rules

import (
	"github.com/goravel/framework/contracts/validation"
	"github.com/goravel/framework/facades"
)

type Unique struct {
}

// Signature The name of the rule.
func (receiver *Unique) Signature() string {
	return "unique"
}

// Passes Determine if the validation rule passes.
func (receiver *Unique) Passes(data validation.Data, val any, options ...any) bool {
	var isExist bool
	var tableName = options[0].(string)
	var columnName = options[1].(string)
	facades.Orm().Query().Table(tableName).Where(columnName, val).Exists(&isExist)

	//

	if isExist {
		return false
	}

	return true
}

// Message Get the validation error message.
func (receiver *Unique) Message() string {
	return ":attribute has been taken."
}
```


Then simply use it like laravel validation for example :
"name": "required|unique:roles,name",

**Comment by xx1196** [comment]
have a rule for validate if a value exist in other table? for example, post have many comments, if i create a comment and i send the post_id, how i can validate this field for not send any value, only acept a id exist in the table post

**Reply by hwbrzzl** [reply]
You can get the `post_id` through `data validation.Data`.

**Comment by haodiao** [comment]
表单是：
 ids[] = 1
 ids[] = 2
怎么验证和获取这个提交的数组的数据

**Reply by hwbrzzl** [reply]
An example: https://github.com/goravel/example/blob/master/app/http/requests/user_create.go#L10-L15

**Comment by xx1196** [comment]
hello, have a question, if im updating 1 resource and have my update form request, if dont send a field this recive in "" how i can ignore null values or empty values for not update if i dont send, but if i send the value but i send in "" this will bee update

**Reply by hwbrzzl** [reply]
Could you write your code here?

**Comment by klphp** [comment]
关于自定义请求验证的问题请教
```
	var formData requests.UserCreateRequest
	var user models.User
	var err error
	
	//目前可以这样使用，但是字段多了手动设置很麻烦，尤其涉及多维数组的时候
	user.Username = formData.Username
	user.Password = formData.Password

	if err = facades.Orm().Query().Create(&user); err != nil {
		return resp.Error(err.Error())
	}

	//有没有更便捷的绑定方法，如
	formData.bind(&user)
	//或
	facades.Orm().Query().Model(&user).Create(&formData)
```

**Reply by hwbrzzl** [reply]
目前可以写一个反射方法，遍历字段去做映射。之后有考虑如何简化这块的操作，但还没有具体方案。

**Comment by tekpriest** [comment]
how can I customize the `errors.All()` to return something like this
```json
{
...
        "email":  "email already exists"
...
    },
```

**Reply by hwbrzzl** [reply]
I think it's very easy to transform it.

**Reply by tekpriest** [reply]
true, I thought there was a way I could that without customizing on response return

**Reply by zerolethanh** [reply]
Hi all, 
Has any idea to set default value of a request? 

```golang 

package requests

import (
	"github.com/goravel/framework/contracts/http"
	"github.com/goravel/framework/contracts/validation"
)

type DotsCmdStoreRequest struct {
	Fname    bool `form:"fname" json:"fname"`
	Download bool `form:"download" json:"download"`
	OutDir   bool `form:"outDir" json:"outDir"`
}

func (r *DotsCmdStoreRequest) Authorize(ctx http.Context) error {
	return nil
}

func (r *DotsCmdStoreRequest) Rules(ctx http.Context) map[string]string {
	return map[string]string{
		"fname":    "bool",
		"download": "bool",
		"outDir":   "bool",
	}
}
```

Now I have to do this to set default value 

```golang 
	var in = requests.DotsCmdStoreRequest{OutDir: false, Fname: true, Download: false}
	_, err := ctx.Request().ValidateRequest(&in)
```

**Reply by krishankumar01** [reply]
@zerolethanh You can tweak the request data in `PrepareForValidation` before validation.
```go
func (r *UserRequest) PrepareForValidation(ctx http.Context, data validation.Data) error {
	if fname, exist := data.Get("fname"); !exist {
		return data.Set("fname", true)
	}

	return nil
}
```

**Reply by zerolethanh** [reply]
thanks you.

**Comment by tekpriest** [comment]
there is a bug when validating nested data, I have this struct
```go
type UpdateTeamPermissionRequest struct {
	Role        string      `form:"role" json:"role"`
	Resource    string      `json:"resource" form:"resource"`
	SubResource SubResource `json:"sub_resource" form:"sub_resource"`
	Actions     []string    `json:"actions" form:"actions"`
	Allowed     bool        `json:"allowed" form:"allowed"`
}
```
and my validation logic
```go
	var data requests.UpdateTeamPermissionRequest
	errors, err := ctx.Request().ValidateRequest(&data)
	if err != nil {
		return ctx.Response().Json(http.StatusBadRequest, c.response.Error(err.Error()))
	}

	if errors != nil {
		return ctx.Response().Json(http.StatusBadRequest, c.response.Error(errors.One()))
	}
```
This is a sample request
```json
{
    "role": "DEVELOPER",
    "resource": "transactions",
    "actions": ["view","export"],
    "sub_resource": {
        "balance": ["manage"]
    }
}
```
The validation passes but when I log it I get this
`{DEVELOPER transactions map[balance:[]] [view export] false}`. The value of balance is empty
but when I log this
```go
	body, _ := io.ReadAll(ctx.Request().Origin().Body)
	fmt.Println("hello", string(body))
```
I get this
```log
hello {
    "role": "DEVELOPER",
    "resource": "transactions",
    "actions": ["view","export"],
    "sub_resource": {
        "balance": ["manage"]
    }
}
```

**Reply by tekpriest** [reply]
this is the rule metadata
```go
func (r *UpdateTeamPermissionRequest) Rules(ctx http.Context) map[string]string {
	return map[string]string{
		"role":         "string",
		"resource":     "string",
		"sub_resource": "map",
		"actions":      "slice",
		"allowed":      "bool",
	}
}
```
I have tried `"sub_resouce.*": "slice", "sub_resource.*": "string"`
all giving same result

**Reply by tekpriest** [reply]
`type SubResource map[string][]string`

**Reply by hwbrzzl** [reply]
Hey @tekpriest Thanks for your feedback. I reproduced it, and I'll release a new version to fix it ASAP. Created an issue: https://github.com/goravel/goravel/issues/533

**Reply by tekpriest** [reply]
superb 👍

**Reply by hwbrzzl** [reply]
Solved in goravel/framework v1.14.9

**Comment by hpwm** [comment]
package middleware

import (
	"github.com/davecgh/go-spew/spew"
	"github.com/goravel/framework/contracts/http"
	"github.com/goravel/framework/facades"
	"github.com/goravel/framework/validation"
	"goravel/app/pkg"
)

type MustHeader struct {
	ProjectUuid string `json:"project_uuid"`
	ClientType  string `json:"client_type"`
	Version     string `json:"version"`
	VersionCode string `json:"version_code"`
	Timestamp   string `json:"timestamp"`
	Signature   string `json:"signature"`
	Nonce       string `json:"nonce"`
}

func Base() http.Middleware {
	return func(ctx http.Context) {
		ctx.WithValue("Request-ID", pkg.BuildUuid(16, false))
		isDebug := ctx.Request().Query("is_debug")
		if isDebug == "xp666wd" {
			ctx.Request().Next()
			return
		}
		headers := map[string]any{
			"project_uuid": ctx.Request().Header("Project-Uuid"),
			"client_type":  ctx.Request().Header("Client-Type"),
			"version":      ctx.Request().Header("Version"),
			"version_code": ctx.Request().Header("Version-Code"),
			"timestamp":    ctx.Request().Header("Timestamp"),
			"signature":    ctx.Request().Header("Signature"),
			"nonce":        ctx.Request().Header("Nonce"),
		}
		validator, err := facades.Validation().Make(headers, map[string]string{
			"project_uuid": "required",
			"client_type":  "required",
			"version":      "required",
			"version_code": "required",
			"timestamp":    "required",
			"signature":    "required",
			"nonce":        "required",
		}, validation.Messages(map[string]string{
			"project_uuid.required": "缺少必要参数1",
			"client_type.required":  "缺少必要参数2",
			"version.required":      "缺少必要参数3",
			"version_code.required": "缺少必要参数4",
			"timestamp.required":    "缺少必要参数5",
			"signature.required":    "缺少必要参数6",
			"nonce.required":        "缺少必要参数7",
		}))
		if err != nil {
			pkg.NewResponse(ctx).Fail(400, err.Error())
			ctx.Request().Abort()
			return
		}
		if validator.Fails() {
			message := validator.Errors().One()
			pkg.NewResponse(ctx).Fail(400, message)
			ctx.Request().Abort()
			return
		}
		var headerArr MustHeader
		validator.Bind(&headerArr)
		spew.Dump(headers, headerArr)
		ctx.Request().Next()
		return
	}
}
为什么headerArr结构体，部分字段没绑定上
不能截图，只能复制
headers:
(map[string]interface {}) (len=7) {
 (string) (len=7) "version": (string) (len=6) "1.21.1",
 (string) (len=12) "version_code": (string) (len=5) "32121",
 (string) (len=9) "timestamp": (string) (len=4) "1334",
 (string) (len=9) "signature": (string) (len=1) "1",
 (string) (len=5) "nonce": (string) (len=3) "122",
 (string) (len=12) "project_uuid": (string) (len=12) "20152020asaa",
 (string) (len=11) "client_type": (string) (len=2) "h5"
}

headerArr
(middleware.MustHeader) {
 ProjectUuid: (string) "",
 ClientType: (string) "",
 Version: (string) (len=6) "1.21.1",
求解
 VersionCode: (string) "",
 Timestamp: (string) (len=4) "1334",
 Signature: (string) (len=1) "1",
 Nonce: (string) (len=3) "122"
}

**Reply by hwbrzzl** [reply]
Sorry, 没看明白，可以提交复现 PR 到 https://github.com/goravel/example 协助处理。

**Comment by wuzhixiang0827** [comment]
请求结构定义
```
type UserStoreRequest struct {
	Name     string `form:"name" json:"name"`
	Account  string `form:"account" json:"account"`
	Password string `form:"password" json:"password"`
	Enabled  int    `form:"enabled" json:"enabled"`
	RoleIds  []any  `form:"role_ids" json:"role_ids"`
}
```

// 验证规则
```
validator, _ := facades.Validation().Make(request, map[string]string{
		"name":     "required",
		"account":  "required",
		"password": "required",
		"enabled":  "required",
		"roleIds":  "required",
	}, validation.Messages(map[string]string{
		"name.required":     "用户名不能为空",
		"account.required":  "账号不能为空",
		"password.required": "密码不能为空",
		"enabled.required":  "状态不能为空",
	}))
```

请求参数
```
{
    "name": "小明",
    "account": "xiaoming",
    "password": "xiaoming",
    "enabled": 1,
    "role_ids": [
        1
    ]
}
```
验证的时候 roleIds是可以的，但要是写成role_ids会提示下面内容，有办法写成role_ids也能正常校验吗？
```
{
	"code": 0,
	"message": "role_ids is required to not be empty"
}
```

**Reply by wuzhixiang0827** [reply]
"roleIds":  "required",  像写成 "role_ids":  "required",

**Reply by wuzhixiang0827** [reply]
"roleIds": "required", 想写成 "role_ids": "required",

**Reply by hwbrzzl** [reply]
你的请求 body 是什么样的？

**Reply by wuzhixiang0827** [reply]
body 是 raw json，json 内容如下：
{
    "name": "小明",
    "account": "xiaoming",
    "password": "xiaoming",
    "enabled": 1,
    "role_ids": [
        1
    ]
}

**Reply by hwbrzzl** [reply]
验证规则中的 key 时根据 body 的 key 设置的，因此 role_ids 应该是可以正确校验的，roleIds 不会通过校验。如果有问题，可以向 https://github.com/goravel/example 提交一个 PR 我看下。

**Comment by damon35868** [comment]
请问有可选值的使用吗？类似于laravel中的表单校验的nullable这种

**Reply by hwbrzzl** [reply]
不加 `required` 即可

**Reply by damon35868** [reply]
好的行了，感谢！

**Reply by soliangD** [reply]
请问下 nullable 字段可以支持为指针nil值吗？测试了下现在拿到的是零值

**Reply by hwbrzzl** [reply]
可以提供下具体代码看下

**Reply by soliangD** [reply]
```go
type OrgCreateRequest struct {
	Test *int `from:"test"`
}

func (r *OrgCreateRequest) Rules(ctx http.Context) map[string]string {
	return map[string]string{
		"test": "int",
	}
}
```
期望传参是null时能获取到nil值，但是好像不支持，获取到的是0,*int

**Reply by soliangD** [reply]
不好意思，这个没问题了，是我这打印逻辑有点问题

**Comment by omiyz** [comment]
string:1,8字符数验证在参数为中文情况下验证过不了，没看源码，仅仅去掉了1,8变成单纯验证string就能通过，猜测是字符数统计方式错了。
ChatGPT:
在 Go 中，中文字符串的长度和英文字符串的长度不一致，原因在于 Go 中的字符串是由 UTF-8 编码表示的，而 UTF-8 编码中的每个字符可能占用不同的字节数。

英文字符通常使用1个字节表示（例如，ASCII字符）。
中文字符通常使用3个字节表示（由于中文字符在 UTF-8 编码下使用 3 个字节来表示）。
因此，虽然 len() 函数会返回字符串的字节长度，但它并不表示字符的数量。对于中文字符和其他多字节字符（如 Emoji），len() 返回的是字节数，而不是字符数。如果你想得到字符串中的“字符数”，而不是字节数，你应该使用 utf8.RuneCountInString() 来计算字符的数量。

**Reply by hwbrzzl** [reply]
目前是在用 [utf8.RuneCountInString()](https://github.com/gookit/validate/blob/v1.5.4/validators.go#L1180)，如果需要可以帖相关代码与请求检查下。

**Comment by omiyz** [comment]
Filters在数组参数格式化时不生效:
``` go

validation.Filters(map[string]string{
		"xxx.*.beginning_date": "str2time",
		"xxx.*.end_date":       "str2time",
		"xxx.*.sign_date":      "str2time",
})

```
参数校验是通过下面方式完成:
```go
facades.Validation().Make(ctx.Request().All(), rules, options...)
```
上传参数如下:
```json
{
  "contracts": [
    {
      "sign_date": "2026-01-04",
      "beginning_date": "2025-01-28",
      "end_date": "2025-07-08"
    },
    {
      "sign_date": "2024-10-04",
      "beginning_date": "2025-09-15",
      "end_date": "2025-10-22"
    }
  ]
}
```
在框架中进行ApplyFilter是其中v.tryGet(field)方法获取到的参数是一个数组:
```go
// gookit工具包里面的方法

// Apply rule for the rule fields
func (r *FilterRule) Apply(v *Validation) (err error) {
	// filter field value
	for _, field := range r.Fields() {
		val, exist, zero := v.tryGet(field)
		if !exist || zero {
			defVal, ok := v.GetDefValue(field)
			// there is also no custom default value
			if !ok {
				continue
			}
........
```
例如上面根据xxx.*.end_date最终获取到的是[]string, 导致str2time根本无法处理, str2time只对string单个时间字符串生效

**Reply by hwbrzzl** [reply]
Sorry for the delay, 目前 github.com/gookit/validate 对 slice 的验证功能还很弱，对于你的例子，目前可以通过自定义 filter 实现：https://www.goravel.dev/zh/the-basics/validation.html#%E8%87%AA%E5%AE%9A%E4%B9%89%E8%BF%87%E6%BB%A4%E8%A7%84%E5%88%99

**Comment by klphp** [comment]
有啥方法能验证 where name=admin and guard=api 是否存在，sql里有复合unique。自定义验证规则好像只能收到当前验证字段的值。

**Reply by hwbrzzl** [reply]
不太明白，可以贴代码看下

**Comment by omiyz** [comment]
- 这个参数校验的规则不能等绑定到结构体之后再进行具体的类型验证吗?   上面的过滤规则int:1,2这种校验再数组参数中根本过不了, 先绑定忽略错误然后再执行validate不是更好?   这个时候结构体的具体类型也有了

**Reply by hwbrzzl** [reply]
验证功能由 https://github.com/gookit/validate 驱动，想法可以贴代码看下，欢迎讨论。 :star:

**Comment by rowjak** [comment]
is there any reference to custom validation language ? like in laravel. example:
lang -> en -> validation.php
return [
    'required' => 'The :attribute field is required.',
    'required_if' => 'The :attribute field is required when :other is :value.',

thanks

**Reply by hwbrzzl** [reply]
We don't support custom validation language for now, an issue can be created if needed.

**Reply by rowjak** [reply]
oke thank you

**Comment by rowjak** [comment]
for max size or min size file ? which validation should i use ? 
thanks

**Reply by krishankumar01** [reply]
Currently, Framework doesn't support file size validation rules out of the box. You can define custom validation rules if needed: https://www.goravel.dev/the-basics/validation.html#custom-validation-rules

You can also keep an eye on issue #693 as we plan to add and improve more validation rules over time.

**Comment by merouanemc** [comment]
Hi,

How can I validate query parameters using a custom Form Request Validation? I tried implementing it, but it doesn’t seem to work as expected. :

**1. URl Format :** http://localhost:3000/records?param1=data&param2=true&param3=22
**2. Form Request Code :**

```go
package requests

import (
	"github.com/goravel/framework/contracts/http"
	"github.com/goravel/framework/contracts/validation"
)

type RecordRequest struct {
	Param1 string `form:"param1" json:"param1"`
	Param2 bool   `form:"param2" json:"param2"`
	Param3 int    `form:"param3" json:"param3"`
}

func (r *RecordRequest) Authorize(ctx http.Context) error {
	return nil
}

func (r *RecordRequest) Rules(ctx http.Context) map[string]string {
	rules := map[string]string{
		"param1": "required|string|max_len:255",
		"param2": "required|bool",
		"param3": "required|int",
	}

	return rules
}

func (r *RecordRequest) Filters(ctx http.Context) map[string]string {
	return map[string]string{}
}

func (r *RecordRequest) Messages(ctx http.Context) map[string]string {
	return map[string]string{}
}

func (r *RecordRequest) Attributes(ctx http.Context) map[string]string {
	return map[string]string{}
}

func (r *RecordRequest) PrepareForValidation(ctx http.Context, data validation.Data) error {

	if param3, exist := data.Get("param3"); exist {
		data.Set("param3", int(param3.(float64)))
	}
	return nil
}
```
**3. Controller code :**
```go
var req requests.RecordRequest
validationErrors, err := ctx.Request().ValidateRequest(&req)
if err != nil {
	return ctx.Response().Json(http.StatusBadRequest, http.Json{
		"message": "Invalid request data",
	  })
}

if validationErrors != nil {
	return ctx.Response().Json(http.StatusUnprocessableEntity, validationErrors.All())
}
```

**4. Error on validation**
```go
interface conversion: interface {} is string, not float64
```

**Reply by krishankumar01** [reply]
This error doesn’t look like a validation failure, but rather it’s caused by this line:
```go
data.Set("param3", int(param3.(float64)))
```
At this point, `param3` is actually a string type, and you’re directly trying to assert it to a `float64`, which will cause a panic.

Instead, you can convert a string to an integer using [filters](https://www.goravel.dev/the-basics/validation.html#available-validation-filters). For example:

```go
func (r *RecordRequest) Filters(ctx http.Context) map[string]string {
	return map[string]string{
		"param3": "toInt",
	}
}

func (r *RecordRequest) PrepareForValidation(ctx http.Context, data validation.Data) error {
	return nil
}
```

**Reply by merouanekhalili** [reply]
I used your suggestion but got the same error for an other params
```go
interface conversion: interface {} is string, not bool
```

Just for your information the Validation works as expected when I send the data within a Post body, but when I send it as Get query params I got this error

**Reply by krishankumar01** [reply]
That’s because for a **POST request body**, the validator uses `json.Unmarshal` into a `map[string]any`.  This preserves the original types.

But for **query parameters**, the data always comes in as `string` values.  It won’t automatically convert them like it does for the request body.

So even though the validator picks up values from query params, they will always be `string`.  To handle this, you need to use [**filters**](https://www.goravel.dev/the-basics/validation.html#available-validation-filters) to explicitly convert those values. (My example was only targeting `param3`. You may need to do the same for `param2` and other parameters as well.)

Ref: internally we use gookit/validate.FromRequest:

https://github.com/gookit/validate/blob/aff789816b9bdbe2287227e7fe7c6a554d3b2295/validate.go#L382
